New Plugin Details page

WBerredo opened the pull request:
### Summary 
Create a Plugin Details page with a new layout as oU6nEIeiKSg8CayhEtgR6b-fi-6559%3A49151
The implementation was made with a new component to be shown when the `marketplace` feature flag is enabled, this component is based on the old `plugin` component, but created as a functional React component.

### Testing instructions
1. Go to plugins page
2. Click in one of the plugins, this should redirect you to the plugin details page.
3. Once on the plugin details page, verify:
- The layout matches with oU6nEIeiKSg8CayhEtgR6b-fi-6559%3A49151
- The mobile version is working as expected
- The CTA actions redirect to a new page when installing and show a thank you page at the end.
- Check the CTA for the 3 possibilities:
  - Install when atomic
  - Upgrade to business and install
  - Install when not atomic

### Screenshots
|Before | After| After Mobile|
|-------|------|------|
|<img width="720" alt="Screen Shot 2021-11-07 at 11 41 09" src="https://user-images.githubusercontent.com/5039531/140652107-985f621d-194e-40ad-adb3-3b578ec8db77.png">|<img width="1188" alt="Screen Shot 2021-11-07 at 11 38 49" src="https://user-images.githubusercontent.com/5039531/140652115-cab3139a-ef5b-483b-9f18-49409743d467.png">|<img width="483" alt="Screen Shot 2021-11-07 at 11 39 42" src="https://user-images.githubusercontent.com/5039531/140652120-18968e17-d3c3-443f-9b1f-b661bf6e88cb.png">|


### Checklist

- [x] Turn a class component into a functional component
- [x] New Layout (wider, 2 columns)
- [x] Header with summary 
- [x] Description / FAQ's
- [x] Include cover image on the description
- [x] Remove the Read more action
- [x] Secondary Info
- [x] Add site list section
- [x] Fix the mobile version
- [x] CTA: Install when atomic
- [x] CTA: Upgrade to business and install
- [x] CTA: Install when not atomic

On separate / next PRs
- [ ] Add placeholder page
- [ ] Add site list styling
- [ ] Add star icons to rating
- [ ] Make sure that author link redirects to our plugin search results showing plugins of that author eg for yoast `https://api.wordpress.org/plugins/info/1.2/?action=query_plugins&request[author]=yoast`

---


Fixes #55577

Commit:
Create a single-plugin functional component to be showed on plugin details page.

This component was based on the class component previously showed.

Commit:
Set a wider width and a two column layout for the single plugin page.

Commit:
Add the header on the Single Plugin page.

Commit:
Change CSS of Single Plugin page body.

cpapazoglou commented on client/my-sites/plugins/controller.js:
```diff
@@ -7,10 +7,10 @@ import getSelectedOrAllSitesWithPlugins from 'calypso/state/selectors/get-select
 import PluginEligibility from './plugin-eligibility';
 import PluginUpload from './plugin-upload';
 import PluginBrowser from './plugins-browser';
+import SinglePluginComponent from './single-plugin';
```
Can we rename this to `PluginDetails` so that it's on par with `PluginUpload`, `PluginBrowser`?

cpapazoglou replied:
Can we circle back to this and rename the component?

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,374 @@
+		recordGoogleEvent,
+		wporgFetchPluginData,
+	}
+)( localize( SinglePlugin ) );
```
Is there a reason for not choosing to use hooks instead of `connect` and `localize`?
Here is an example 
https://github.com/Automattic/wp-calypso/blob/1ead2e43c770d79fa65b32b8135b393a434369a9/client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx#L26-L45

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,374 @@
+		}
+
+		return !! props.sitePlugin;
+	}, [ requestingPluginsForSites, props.sitePlugin ] );
```
Do we need to use `useMemo` here? I _feel_ that these computations are fairly simple.

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,374 @@
+			) }
+		</div>
+	);
+}
```
How would you feel splitting this distinct sections into different components used by the current one?

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,374 @@
+	const recordBackArrowClickedEvent = recordEvent( 'Clicked Header Plugin Back Arrow' );
+	const shouldUseHistoryBack = window.history.length > 1 && props.navigated;
+	return (
+		<HeaderCake
```
How would you feel splitting this distinct sections into different components used by the current one?

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,374 @@
+		return false;
+	}, [ isFetching, isFetched, fullPlugin, requestingPluginsForSites ] );
+
+	function getPageTitle() {
```
nitpick: let's follow the fun "arrow function" pattern when declaring new functions.
```suggestion
	const getPageTitle = () => {
```

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,374 @@
+	}
+
+	const isWpcom = selectedSite && ! props.isJetpackSite;
+	const analyticsPath = selectedSite ? '/plugins/:plugin/:site' : '/plugins/:plugin';
```
can we move these statements to the top? or at least above the return early statements?

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,374 @@
+}
+
+function Header( props ) {
+	if ( ! props.selectedSite ) {
```
Probably this should be replaced with the new [Fixed Header](https://github.com/Automattic/wp-calypso/pull/57392), and will be enhanced with https://github.com/Automattic/wp-calypso/pull/57442

cpapazoglou commented on client/my-sites/plugins/style.scss:
```diff
@@ -93,3 +95,118 @@
+
+.layout__content {
+	background: var( --studio-white );
+}
```
This is pretty generic and also creates the following
![SS 2021-11-02 at 12 53 36](https://user-images.githubusercontent.com/12430020/139833711-6f338bd5-7565-42ce-ae57-7b4245b472ff.png)

Basically this
https://github.com/Automattic/wp-calypso/pull/57225/files#diff-69e8e55283f629d82fde8f7811715de63d5a865003eaa1c8e775ee8e6eab824cR13-R14

will most likely solve the problem. Would you mind rebasing?

cpapazoglou reviewed (commented):
I have been doing some premature review on this so if you were going to address any of my comments, just ignore them 😅.

BTW, it looks good!

Let's have in mind the following scenarios for the CTA:

1. All plugins page `/plugins`, should not be showing that CTA but something like 
![SS 2021-11-02 at 13 05 39](https://user-images.githubusercontent.com/12430020/139835380-b8d0cc53-18e0-48d9-98ac-7c9d083ef981.png)
2. Simple Site with lower than business plan, should read "Upgrade and install"
3. Simple / Atomic site with business or ecommerce plan, should read "Install and activate"

Commit:
Add feature flag check for the Plugin Details page.

Commit:
Add additional information to the right column of the Plugin Details page.

Commit:
Add the addBanner and removeReadMore on the Plugin Sections component to be used on the Plugin Details page.

Commit:
Small fixes amd lint issues.

Commit:
Update the Single Plugin page to use hooks.

Commit:
Update layout and texting of CTA.

Commit:
Add CTA for install and activate plugin.

Commit:
Plugin Details: adds header navigation

Commit:
Fixes multisite.

Commit:
Hide image component when no banner is given

Commit:
Add Plugin List to Plugin details page and mobile layout updates.

Commit:
Plugin Details: bootstrap flow for checking out and installing plugin.

Commit:
Add CTA to Plugin Details page.

cpapazoglou commented on client/state/plugins/installed/actions.js:
```diff
@@ -438,6 +443,10 @@ export function installPluginOnMultisite( siteId, plugin ) {
 	return installPluginHelper( siteId, plugin, true );
 }
 
+export function installPluginSimple( siteId, plugin ) {
+	return installPluginHelper( siteId, plugin, false, true );
+}
+
```
It seems that `installPluginSimple` is avoiding plugin activation which is done through `activatePlugin`, can we instead change `installPlugin( siteId, plugin )` to accept a `activatePlugin` parameter? `simpleInstallation` is not descriptive enough and also devs may confuse it with Simple vs Atomic.

WBerredo replied:
Agreed, I did take some time thinking about the name, being explicit about the action seems like the best way here.

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -322,7 +302,7 @@ function CTA( { slug, isPluginInstalledOnsite, fullPlugin } ) {
 			<Button
 				className="single-plugin__install-button"
-				onClick={ () => onClickInstallPlugin( { dispatch, selectedSiteId, fullPlugin, slug } ) }
+				href={ `/marketplace/${ slug }/install/${ selectedSite.slug }` }
```
We will need to use `onClick` both here and in https://github.com/Automattic/wp-calypso/pull/57478/commits/f10698fd4e9b34a714857b83bc1491595dbe0a20#diff-fbb59cfe300fa8fdfab594360525688a332b9319de457754a6b79f4266bf8d2cR315 so that we still dispatch the "click" track events.

```
dispatch( recordGoogleEvent( 'Plugins', 'Install on selected Site', 'Plugin Name', slug ) );
	dispatch(
		recordGoogleEvent( 'calypso_plugin_install_click_from_plugin_info', {
			site: selectedSiteId,
			plugin: slug,
		} )
	);
```

WBerredo replied:
Cool, will make generalize the code for both options and send the events.

cpapazoglou commented on client/my-sites/marketplace/index.js:
```diff
@@ -34,9 +34,16 @@ export default function () {
 		);
 		page(
-			'/marketplace/product/install/:site',
+			'/marketplace/product/install/:site?',
```
Do we need this route? It should be covered by `'/marketplace/:productSlug?/install/:site?'` which also accepts the `:productSlug` parameter as optional.

cpapazoglou replied:
Changing 

```
page( `/marketplace/product/install/${ nextProps.siteSlug }` );
```
to 
```
page( `/marketplace/install/${ nextProps.siteSlug }` );
```
here https://github.com/Automattic/wp-calypso/blob/1ead2e43c770d79fa65b32b8135b393a434369a9/client/my-sites/plugins/plugin-upload/index.jsx#L61  should be enough to be picked up by 
```
'/marketplace/:productSlug?/install/:site?'
```

WBerredo replied:
Applied!

Commit:
Small fixes about variable naming, URL naming and readding tracking events.

Commit:
Use the redirect method when finishing the plugin installation.

This change prevents the user from go to the progress page once in thank you page.

Commit:
Add the trasfer to atomic flow.

Commit:
Rename SinglePlugin to PluginDetails

cpapazoglou commented on client/my-sites/plugins/style.scss:
```diff
@@ -110,3 +116,186 @@ body.is-section-plugins:not( .is-nav-unification ) {
+			border-bottom: 1px solid var( --studio-gray-5 );
+		}
+	
+		.single-plugin__version {
```
Let's add the version in the header too

![SS 2021-11-09 at 18 17 12](https://user-images.githubusercontent.com/12430020/140962393-a6681aa5-313e-4cc9-bc27-027dbbe0d8f2.png)

WBerredo replied:
Sure, added.

cpapazoglou commented on client/my-sites/plugins/single-plugin.jsx:
```diff
@@ -0,0 +1,438 @@
+							),
+						} ) }
+					>
+						<div className="single-plugin__header">
```
Can we add the short description and the tags in the header too?
![SS 2021-11-09 at 18 18 56](https://user-images.githubusercontent.com/12430020/140963121-93622065-ffca-4960-83fd-c2235f347546.png)

WBerredo replied:
> Can we add the short description and the tags in the header too?

Checking the disponibility of those informations

> padding

I will work on it now

> stats( developer, ratings) should be 2 column. I am not sure why you used table there to be honest. Is there something that can't be achieved with flexbox?

No strong reason to be using table, I just tough it would make more sense semantically, but considering the mobile version I will use flexbox.

Commit:
Rename SinglePlugin to PluginDetails and add version to the header.

cpapazoglou reviewed (changes requested):
**In Mobile we need to make some adjustments in the header:**
1. paddings
2. stats( developer, ratings) should be 2 column. I am not sure why you used table there to be honest. Is there something that can't be achieved with flexbox?

<img width="914" alt="SS 2021-11-09 at 18 24 03" src="https://user-images.githubusercontent.com/12430020/140964106-42c043d0-b180-461a-aa8f-97bf956767c1.png">

**After installing contact form I got** 

![image](https://user-images.githubusercontent.com/12430020/140966199-7c2ea395-a1be-484a-9154-362d616773f7.png)

it’s the plugin slug missing which is causing the problems

cpapazoglou commented:
I think it would be better reserving the right space on the header while waiting for the CTA so that we don't have a layout shift. In a future iteration we can also add a loader.

![SS 2021-11-09 at 18 48 47](https://user-images.githubusercontent.com/12430020/140968031-aa89bc05-2a08-4109-a54b-787c19ada12a.gif)

Commit:
Fix plugin slug addition to thank you page url.

WBerredo commented:
> I think it would be better reserving the right space on the header while waiting for the CTA

Sure, but I will probably need to check if it exists and doesn't show on mobile. Otherwise we would be showing a blank space below the header.

Commit:
Plugin details: adds short_description.

cpapazoglou commented:
> > I think it would be better reserving the right space on the header while waiting for the CTA
> 
> Sure, but I will probably need to check if it exists and doesn't show on mobile. Otherwise we would be showing a blank space below the header.

In mobile we either show the CTA or the site list in the same space.

cpapazoglou commented on client/my-sites/plugins/style.scss:
```diff
@@ -110,3 +116,183 @@ body.is-section-plugins:not( .is-nav-unification ) {
+	height: 400px;
+
+	.plugin-details__name, .plugin-details__price {
+		font-family: Recoleta;
```
Let's use `$brand-serif` instead of `Recoleta`
```suggestion
		font-family: $brand-serif;
```

cpapazoglou commented on client/my-sites/plugins/plugin-details.jsx:
```diff
@@ -0,0 +1,440 @@
+	const selectedSite = useSelector( getSelectedSite );
+	const selectedSiteId = useSelector( getSelectedSiteId );
+
+	if ( ! shouldDisplayCTA( selectedSite, slug, isPluginInstalledOnsite ) ) {
+		return null;
+	}
+
+	const shouldUpgrade = ! (
+		isBusiness( selectedSite.plan ) ||
+		isEnterprise( selectedSite.plan ) ||
+		isEcommerce( selectedSite.plan )
+	);
```
Jetpack self hosted sites shouldn't need to upgrade - buy a business plan.

cpapazoglou replied:
This is an example of how we can figure out a Jetpack self hosted site
`https://github.com/Automattic/wp-calypso/blob/038f3877a3e639dceda8ba5e5dcbf6bc18d94632/client/my-sites/plugins/plugins-browser/index.jsx#L567`

WBerredo replied:
For what I understood if a site is already jetpack they won't need to upgrade. Updating the code to:

```ts
const isJetpack = useSelector( ( state ) => checkJetpackSite( state, selectedSite.ID ) );
const shouldUpgrade = ! (
    isBusiness( selectedSite.plan ) ||
    isEnterprise( selectedSite.plan ) ||
    isEcommerce( selectedSite.plan ) ||
    isJetpack
);
```

cpapazoglou replied:
Seems good 👌

cpapazoglou commented on client/my-sites/plugins/plugin-details.jsx:
```diff
@@ -0,0 +1,440 @@
+	const dispatch = useDispatch();
+	const translate = useTranslate();
+	const selectedSite = useSelector( getSelectedSite );
+	const selectedSiteId = useSelector( getSelectedSiteId );
```
since we already requested `selectedSite` we can directly use `selectedSite.ID` instead of fetching it again from the state.

cpapazoglou commented on client/my-sites/plugins/plugin-details.jsx:
```diff
@@ -0,0 +1,440 @@
+		if ( ! isFetched ) {
+			dispatch( wporgFetchPluginData( props.pluginSlug ) );
+		}
+	}, [ isFetched ] );
```
is there a reason for not adding `dispatch` and `props.pluginSlug` as dependencies?

WBerredo replied:
It was a quick migration from the class component, as you can see [here](https://github.com/Automattic/wp-calypso/blob/5a88fc6a29779919968b6e2f4e469f6e2fccc210/client/my-sites/plugins/plugin.jsx#L47-L52).

That said, I'm fixing on the next commit.

cpapazoglou commented on client/my-sites/plugins/plugin-details.jsx:
```diff
@@ -0,0 +1,440 @@
+	const moment = useLocalizedMoment();
+	const translate = useTranslate();
+
+	const selectedSiteId = useSelector( getSelectedSiteId );
```
we can use `selectedSite.ID` directly since we 've got `const selectedSite = useSelector( getSelectedSite );`

Commit:
Small layout fixes.

cpapazoglou commented on client/my-sites/marketplace/pages/marketplace-plugin-upload-status/index.tsx:
```diff
@@ -54,29 +72,66 @@ const MarketplacePluginUpload = (): JSX.Element => {
+			transferStates.COMPLETE === automatedTransferStatus
+		) {
+			setCurrentStep( 2 );
+		}
+	}, [ moveToAtomicFlow, automatedTransferStatus ] );
+
+	useEffect( () => {
+		if (
+			installedPlugin &&
+			currentStep === 1 &&
+			( ! isUploadFlow || ( isUploadFlow && pluginUploadComplete ) )
+		) {
 			dispatch(
-				togglePluginActivation( siteId, {
-					slug: uploadedPlugin?.slug,
-					id: uploadedPlugin?.id,
+				activatePlugin( siteId, {
+					slug: installedPlugin?.slug,
+					id: installedPlugin?.id,
 				} )
 			);
 			setCurrentStep( 2 );
 		}
 		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [ pluginUploadComplete, uploadedPlugin, setCurrentStep ] );
+	}, [ pluginUploadComplete, installedPlugin, setCurrentStep ] );
 
 	useEffect( () => {
-		if ( pluginActive ) {
+		if (
+			( installedPlugin && pluginActive ) ||
+			( moveToAtomicFlow && transferStates.COMPLETE === automatedTransferStatus )
+		) {
 			waitFor( 1 ).then( () =>
-				page( `/marketplace/thank-you/${ uploadedPlugin?.slug }/${ selectedSiteSlug }` )
+				page.redirect(
+					`/marketplace/thank-you/${ installedPlugin?.slug || productSlug }/${ selectedSiteSlug }`
+				)
 			);
 		}
```
Since we added the install flow here together with the upload flow, can we add a small comment to each `userEffect` to signal why they are used?

Commit:
Update the HTML semantic of the additional information section to use flexbox instead of table.

Commit:
Consider Jetpack when showing the upgrade button and usage of selectedSite.ID instead of retrieve ID from state all time.

Commit:
Add comments to markplace plugin status page.

Commit:
Add required props to a useEffect.

Commit:
Plugin Details: Optionally chains site id.

Commit:
Plugin Details: clears up PluginSiteList.

Commit:
Plugin install: fixes grey bg for users using the upgrade flow.

Commit:
Plugin Details: simplifies selectors which use optional chaining.

cpapazoglou reviewed (approved):
Thanks for addressing all feedback @WBerredo! I have tested the following flows:

- free site -> upgrades to business, transfers to atomic, and activates plugin
- business simple -> transfers to atomic, and activates plugin
- atomic site -> activates plugin
- jetpack self hosted -> activates plugin

We can merge this as it's behind a feature flag and iterate on it!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6955190

Thank you @WBerredo for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6955190).
